### PR TITLE
CurlHandler: Remove Transfer-Encoding on redirect from POST to GET

### PR DESCRIPTION
**Pull request by rmkerr**
This fix removes the TransferEncoding header from a request when it is redirected from POST to GET. I chose to modify the request for the following reasons:

1. It's consistent with the behavior of WinHttpHandler.
2. At this point the header has already been modified by the code linked below.

https://github.com/dotnet/corefx/blob/444e6bb7fd274be680fb3bf4b22258821e77ae2c/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.cs#L777-L781

I also tagged the locations where the TransferEncoding header is being modified with an issue tracking removal of those modifications.

If anyone feels strongly about not modifying the request (despite the above), I think that it is possible. It's just a much more significant rework.

Fixes: #27301

### CODE REVIEW

**Comment by davidsh** [comment]
>It's consistent with the behavior of WinHttpHandler.

Where does WinHttpHandler do that? I'm not aware that WinHttpHandler will modify the original HttpRequestMessage.Method when redirecting from POST to GET.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs [hunk 1]
```diff
@@ -999,6 +999,12 @@ private static ulong CurlReceiveHeadersCallback(IntPtr buffer, ulong size, ulong
                                     // our redirect limit, etc.), this will have been unnecessary work in reconfiguring the easy handle, but 
                                     // nothing incorrect, as we'll tear down the handle once the request finishes, anyway, and all of the configuration
                                     // we're doing is about initiating a new request.
+                                    if ((int)response.StatusCode >= 301 && (int)response.StatusCode <= 303)
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: why use (int) here?  Why not just use the HttpStatusCode enum type directly?

**Comment by rmkerr** [comment]
WinHtppHandler doesn't modify the `HttpRequestMessage.Method`, but it does modify `Transfer-Encoding` and `Content-Length`.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs (hunk 1 above)
Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: I based this on the lines just above this one:
https://github.com/dotnet/corefx/blob/444e6bb7fd274be680fb3bf4b22258821e77ae2c/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs#L991-L993
But I can always adjust to use `HttpStatusCode`. That would definitely be more clear.

**Comment by davidsh** [comment]
>WinHtppHandler doesn't modify the HttpRequestMessage.Method, but it does modify Transfer-Encoding and Content-Length.

Sorry, I meant to imply "WinHttpHandler will not modify Transfer-Encoding headers when redirecting from POST to GET.".  You are right that in other situations (like conflict between the headers or neither header set) WinHttpHandler will adjust original headers in those cases.

Given your assessment that fixing this completely appears to be complex, I agree that the quick fix to functionality is what you have in this PR.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs (hunk 1 above)
Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: I don't think that is a best-practice pattern either.  The StatusCode property already has a type associated with it (HttpStatusCode). So, the best thing is just leave it as that type and use the enum values to compare.

Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: Sounds good. I'll adjust my change and update the lines above as well.

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: >update the lines above as well.

The lines above which are testing for multiple values in a range might be ok to use `int` casts. But for checking individual values leaving it as HttpStatusCode is best.

Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: My code is also checking for a range of values. I could change it to check each of the three relevant status codes, but the current code makes fewer comparisons.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 2]
```diff
@@ -641,11 +641,10 @@ public void GetAsync_ServerNeedsAuthAndNoCredential_StatusCodeUnauthorized()
         [InlineData(303)]
         public async Task AllowAutoRedirect_True_PostToGetDoesNotSendTE(int statusCode)
         {
-            if (IsCurlHandler)
+            if (IsCurlHandler && statusCode == 300)
```
Comment (rmkerr) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: CURL does not offer the option to redirect from POST to GET on a 300 multiple choices redirect. It's actually a little unclear if this behavior is incorrect, but it has been discussed before. See #26434 for more details.

**Comment by rmkerr** [comment]
@dotnet-bot test outerloop Windows x64 Debug Build
@dotnet-bot test outerloop linux x64 Debug Build
@dotnet-bot test outerloop OSX x64 Debug Build

**Comment by rmkerr** [comment]
Hey @geoffkizer this is resolving an issue you found -- mind taking a look to see if this is in line with what you expected?

**Comment by geoffkizer** [comment]
LGTM

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs [hunk 3]
```diff
@@ -999,6 +999,12 @@ private static ulong CurlReceiveHeadersCallback(IntPtr buffer, ulong size, ulong
...
+                                    {
+                                        // ISSUE: 25163
+                                        // Ideally we want to avoid modifying the users request message.
+                                        easy._requestMessage.Headers.TransferEncodingChunked = false;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: Is `false` what we want, or would `null` be better?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: libcurl appears to be following the spec, https://tools.ietf.org/html/rfc7231#section-6.4.3.  For 301 and 302, it states "Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request", and for 303 it states "The 303 (See Other) status code indicates that the server is redirecting the user agent to a different resource, as indicated by a URI in the Location header field, which is intended to provide an indirect response to the original request.  A user agent can perform a retrieval request targeting that URI (a GET or HEAD request if using HTTP)".  But unless I missed it, the description of 300 doesn't say anything that suggests it should result in changing POST to GET. curl's code for this is here: https://github.com/curl/curl/blob/4a9f14d921d1f32831aa8956294db1a74a4dc459/lib/transfer.c#L1826.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs (hunk 3 above)
Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs]: I think `false` is the right choice, based off the code that blocks the header:
https://github.com/dotnet/corefx/blob/52bba5248ba4105140f8f62359da7779d8860d0b/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L732-L738
It's also what `CurlHandler` does alsewhere:
https://github.com/dotnet/corefx/blob/444e6bb7fd274be680fb3bf4b22258821e77ae2c/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.cs#L777-L781

**Comment by rmkerr** [comment]
The failures in CI are already tracked by #26092 and #25878. There are also failures in `System.ComponentModel.Composition.Tests` and `System.Reflection.Metadata.Tests`, but the failures are not relevant to the changes here.

**Pull request merged**